src/fastfield/multivalued/writer.rs

```diff
@@ -200,24 +184,119 @@ impl MultiValuedFastFieldWriter {
                             .iter()
                             .map(|val| *mapping.get(val).expect("Missing term ordinal"));
                         for val in remapped_vals {
-                            value_serializer.add_val(val)?;
+                            values.push(val);
                         }
                     }
                 }
             } else {
-                let val_min_max = minmax(self.vals.iter().cloned());
-                let (val_min, val_max) = val_min_max.unwrap_or((0u64, 0u64));
-                value_serializer =
-                    serializer.new_u64_fast_field_with_idx(self.field, val_min, val_max, 1)?;
                 for vals in self.get_ordered_values(doc_id_map) {
                     // sort values in case of remapped doc_ids?
                     for &val in vals {
-                        value_serializer.add_val(val)?;
+                        values.push(val);
                     }
                 }
             }
-            value_serializer.close_field()?;
+            let col = VecColumn::from(&values[..]);
+            serializer.create_auto_detect_u64_fast_field_with_idx(self.field, col, 1)?;
         }
         Ok(())
     }
 }
+
+pub(crate) struct MultivalueStartIndex<'a, C: Column> {
+    column: C,
+    doc_id_map: &'a DocIdMapping,
+    min_max_opt: Mutex<Option<(u64, u64)>>,
+}
+
+impl<'a, C: Column> MultivalueStartIndex<'a, C> {
+    pub fn new(column: C, doc_id_map: &'a DocIdMapping) -> Self {
+        assert_eq!(column.num_vals(), doc_id_map.num_old_doc_ids() as u64 + 1);
+        MultivalueStartIndex {
+            column,
+            doc_id_map,
+            min_max_opt: Mutex::default(),
+        }
+    }
+
+    fn minmax(&self) -> (u64, u64) {
+        if let Some((min, max)) = self.min_max_opt.lock().unwrap().clone() {
+            return (min, max);
+        }
+        let (min, max) = tantivy_bitpacker::minmax(self.iter()).unwrap_or((0u64, 0u64));
+        *self.min_max_opt.lock().unwrap() = Some((min, max));
+        (min, max)
+    }
+}
+impl<'a, C: Column> Column for MultivalueStartIndex<'a, C> {
+    fn get_val(&self, idx: u64) -> u64 {
+        self.iter().nth(idx as usize).unwrap()
+    }
+
+    fn min_value(&self) -> u64 {
+        self.minmax().0
+    }
+
+    fn max_value(&self) -> u64 {
+        self.minmax().1
+    }
+
+    fn num_vals(&self) -> u64 {
+        (self.doc_id_map.num_new_doc_ids() + 1) as u64
+    }
+
+    fn iter<'b>(&'b self) -> Box<dyn Iterator<Item = u64> + 'b> {
+        Box::new(MultivalueStartIndexIter {
+            column: &self.column,
+            doc_id_map: self.doc_id_map,
+            new_doc_id: 0,
+            offset: 0,
+        })
+    }
```

Comment:
We can use iter map here
```suggestion
        let mut offset: u64 = 0;
        Box::new(
            std::iter::once(0_u64).chain(self.doc_id_map.iter_old_doc_ids().map(
                move |old_docid| {
                    let start_pos = self.column.get_val(old_docid as u64);
                    let end_pos = if old_docid as u64 + 1 >= self.column.num_vals() {
                        self.column.num_vals() as u64
                    } else {
                        self.column.get_val(old_docid as u64 + 1)
                    };
                    offset += end_pos - start_pos;
                    offset
                },
            )),
        )

```